source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.cc

```diff
@@ -0,0 +1,63 @@
+#include "extensions/http/header_formatters/preserve_case/preserve_case_formatter.h"
+
+#include "envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.pb.h"
+#include "envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.pb.validate.h"
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderFormatters {
+namespace PreserveCase {
+
+std::string PreserveCaseHeaderFormatter::format(absl::string_view key) const {
+  const auto remembered_key = original_header_keys_.find(key);
+  // TODO(mattklein123): We can avoid string copies here if the formatter interface allowed us
+  // to return something like GetAllOfHeaderAsStringResult with both a string_view and an
+  // optional backing string. We can do this in a follow up if there is interest.
+  if (remembered_key != original_header_keys_.end()) {
+    return *remembered_key;
+  } else {
+    return std::string(key);
+  }
+}
+
+void PreserveCaseHeaderFormatter::rememberOriginalHeaderKey(absl::string_view key) {
+  // Note: This implementation will only remember the first instance of a particular header key.
+  // So for example "Foo" followed by "foo" will both be serialized as "Foo" on the way out. We
+  // could do better here but it's unlikely it's worth it and we can see if anyone complains about
+  // the implementation.
```

Comment:
It's also unclear how that would work right when headers are modified: what should happen if upstream responds with 4 headers with the same key but different casing and one filter removes the second occurrence and adds another value? +1 to deferring this until someone asks for ti